### PR TITLE
[CMake] Use CMAKE_MSVC_RUNTIME_LIBRARY to apply /MT

### PR DESCRIPTION
**Pull request by hcho3**
Use the built-in flag `CMAKE_MSVC_RUNTIME_LIBRARY` in CMake to robustly handle the linking mode for the Windows C runtime (`/MD` vs `/MT`). See https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -207,6 +207,11 @@ if(__INTEGRATE_OPENCL)
     endif()
 endif()
 
+if(BUILD_CPP_TEST AND MSVC)
+  # Use /MT flag to statically link the C runtime
+  set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
+endif()
```
Comment (hcho3) [inline CMakeLists.txt:213]: Not sure what the default should be for LightGBM when it's not built with gtest. For now, I kept the current behavior, which is to specify `/MT` when gtest is enabled.

Comment (jameslamb) [reply CMakeLists.txt:213]: I agree with this, let's keep the current behavior.

And for now, could you please put this down where the existing string-replacement code is? For as long as the testing context is the only place where we explicitly set this, I'd prefer to keep this inside the big `if(BUILD_CPP_TEST)` block, alongside most of the other test-specific configuration.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Looks good to me! Just left one small request to move this code back down inside the `if(BUILD_CPP_TEST)` block.

Don't worry about the failing CI jobs... they're flaky and the failures are unrelated to your changes.

File: CMakeLists.txt [hunk 2]
```diff
@@ -679,18 +679,8 @@ if(BUILD_CPP_TEST)
...
-        CMAKE_CXX_FLAGS_RELEASE
-        CMAKE_C_FLAGS
-        CMAKE_C_FLAGS_DEBUG
-        CMAKE_C_FLAGS_RELEASE
-    )
-    foreach(CompilerFlag ${CompilerFlags})
-      string(REPLACE "/MD" "/MT" ${CompilerFlag} "${${CompilerFlag}}")
-    endforeach()
+    # Use /MT flag to statically link the C runtime
+    set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
```
Comment (jameslamb) [inline CMakeLists.txt:683]: @hcho3 could you please look into https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=16422&view=logs&j=1f4df553-f999-5fff-c6fe-71123c872ae2&t=3537c7a9-e1e2-54d5-18f4-f7d9e02b9f6d?

```text
boosting.obj : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MDd_DynamicDebug' doesn't match value 'MTd_StaticDebug' in test_array_args.obj [D:\a\1\s\build\testlightgbm.vcxproj]
```

Does this maybe need to be set earlier, before the target is declared, to affect the flags? I'm not really sure.

Comment (hcho3) [reply CMakeLists.txt:683]: Yes, I think this has to be set befor3 targets are created.

**Review by jameslamb — APPROVED** [review]
thanks very much @hcho3 ! This is a nice improvement. I'm glad to be able to get rid of some string manipulation in `CMakeLists` 😁

**Pull request merged**